File: tcp_cong/tcp_cong.py
#!/usr/bin/env python3

"""
tcp_cong - TCP Congestion Tracker

You can find what means each state and more information in here:
https://github.com/torvalds/linux/blob/master/include/uapi/linux/tcp.h#L194

ToDo:
- Add specific pid and comm
"""

from bcc import BPF
import socket
import struct

b = BPF(src_file="tcp_cong.c")

print(f"{'PID':<8} {'COMM':<16} {'DADDR':<15} {'DPORT':<10} {'CA_STATE':<10} {'CWND':<10}")
print("-" * 68)

def print_event(cpu, data, size):
    event = b["events"].event(data)
    d_ip_address = socket.inet_ntoa(struct.pack("=I", event.daddr))

    print((
        f"{event.pid:<8} {event.comm.decode('utf-8', 'replace'):<16} "
        f"{d_ip_address:<15} {event.dport:<10} {event.ca_state:<10} {event.cwnd:<10}"
    ))

b["events"].open_perf_buffer(print_event)

try:
    while True:
        b.perf_buffer_poll()
except KeyboardInterrupt:
    print("\nTschüss!")

File: tcp_cong/tcp_cong.c
#include <uapi/linux/ptrace.h>
#include <net/sock.h>
#include <linux/tcp.h>
#include <linux/bpf.h>

struct output_t {
  u32 pid;
  char comm[TASK_COMM_LEN];
  u32 daddr;
  u16 dport;
  u8 ca_state;
  u32 cwnd;
};

BPF_PERF_OUTPUT(events);

int kprobe__tcp_set_ca_state(struct pt_regs *ctx, struct sock *sk, const u8 ca_state) {
  struct output_t output = {};
  output.pid = bpf_get_current_pid_tgid() >> 32;
  bpf_get_current_comm(&output.comm, sizeof(output.comm));
  u16 dport = sk->__sk_common.skc_dport;
  output.dport = ntohs(dport);
  output.daddr = sk->__sk_common.skc_daddr;
  output.ca_state = ca_state;
  struct tcp_sock *tp = (struct tcp_sock *)sk;
  output.cwnd = tp->snd_cwnd;
  events.perf_submit(ctx, &output, sizeof(output));
  return 0;
}
